compiler/ast.rs

```diff
@@ -0,0 +1,633 @@
+          );
+        }
+
+        msg
+      })
+      .collect::<Vec<String>>();
+
+    Self(diagnostics)
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct ErrorBuffer(Arc<RwLock<Vec<Diagnostic>>>);
+
+impl ErrorBuffer {
+  pub fn new() -> Self {
+    Self(Arc::new(RwLock::new(Vec::new())))
+  }
+}
+
+impl Emitter for ErrorBuffer {
+  fn emit(&mut self, db: &DiagnosticBuilder) {
+    self.0.write().unwrap().push((**db).clone());
+  }
+}
+
+fn get_es_config(jsx: bool) -> EsConfig {
+  EsConfig {
+    class_private_methods: true,
+    class_private_props: true,
+    class_props: true,
+    dynamic_import: true,
+    export_default_from: true,
+    export_namespace_from: true,
+    import_meta: true,
+    jsx,
+    nullish_coalescing: true,
+    num_sep: true,
+    optional_chaining: true,
+    top_level_await: true,
+    ..EsConfig::default()
+  }
+}
+
+fn get_ts_config(tsx: bool) -> TsConfig {
+  TsConfig {
+    decorators: true,
+    dynamic_import: true,
+    tsx,
+    ..TsConfig::default()
+  }
+}
+
+fn get_syntax(media_type: MediaType) -> Syntax {
+  match media_type {
+    MediaType::JavaScript => Syntax::Es(get_es_config(false)),
+    MediaType::JSX => Syntax::Es(get_es_config(true)),
+    MediaType::TypeScript => Syntax::Typescript(get_ts_config(false)),
+    MediaType::TSX => Syntax::Typescript(get_ts_config(true)),
+    _ => Syntax::Es(get_es_config(false)),
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct TranspileOptions {
+  emit_metadata: bool,
+  jsx_factory: String,
+  jsx_factory_fragment: String,
+  transform_jsx: bool,
+}
+
+impl Default for TranspileOptions {
+  fn default() -> Self {
+    TranspileOptions {
+      emit_metadata: false,
+      jsx_factory: "React.createElement".into(),
+      jsx_factory_fragment: "React.Fragment".into(),
+      transform_jsx: true,
+    }
+  }
+}
+
+#[derive(Clone)]
+pub struct ParsedModule {
+  comments: SingleThreadedComments,
+  module: Module,
+  source_map: Rc<SourceMap>,
+}
+
+impl fmt::Debug for ParsedModule {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "ParsedModule {{ }}")
+  }
+}
+
+impl ParsedModule {
+  pub fn get_leading_comments(&self, span: &Span) -> Vec<Comment> {
+    self
+      .comments
+      .with_leading(span.lo, |comments| comments.to_vec())
+  }
+
+  /// Identifies all syntactical dependencies of a module (e.g. `import` and
+  /// `export` statements).
+  pub fn get_dependencies(&self) -> Vec<DependencyDescriptor> {
+    let mut collector = DependencyCollector(Vec::new());
+    collector.visit_module(&self.module, &self.module);
+
+    collector.0
+  }
+
+  pub fn get_location(&self, span: &Span) -> Location {
+    self.source_map.lookup_char_pos(span.lo).into()
+  }
+
+  pub fn get_module_leading_comments(&self) -> Vec<Comment> {
+    self.get_leading_comments(&self.module.span)
+  }
+
+  pub fn transpile(&self, options: TranspileOptions) -> Result<String> {
+    let program = Program::Module(self.module.clone());
```

Comment:
Watch out, cloning `Module` AST is not a good idea, we've already had big performance impact in `deno_lint` when doing so. Maybe this method could consume `self` instead of taking a reference?

Reply:
👍